crates/matrix-sdk-ui/src/app/mod.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for that specific language governing permissions and
+// limitations under the License.
+
+//! Unified API for both the Room List API and the Encryption Sync API, that takes care of all the
+//! underlying details. This is an opiniated way to run both APIs, with high-level callbacks that
+//! should be called in reaction to user actions and/or system events.
+
+use std::sync::{Arc, Mutex};
+
+use eyeball::shared::Observable;
+use futures_core::Stream;
+use futures_util::{pin_mut, StreamExt as _};
+use matrix_sdk::Client;
+use thiserror::Error;
+use tokio::task::{spawn, JoinHandle};
+
+use crate::{
+    encryption_sync::{self, EncryptionSync, WithLocking},
+    room_list::{self, RoomListService},
+};
+
+#[derive(Clone)]
+pub enum AppState {
+    Running,
+    Terminated,
+    Error,
+}
+
+pub struct App {
+    room_list: Arc<RoomListService>,
+    encryption_sync: Option<Arc<EncryptionSync>>,
```

Comment:
If we're always going to have an `EncryptionSync` then we don't need an optional here. We could `#[cfg(feature = "e2e-encryption")]` this field if we want to support clients that don't use E2EE.